src/gui/src/heatMapPinDensity.cpp

```diff
@@ -0,0 +1,175 @@
+//////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (c) 2024, Precision Innovations Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "heatMapPinDensity.h"
+
+#include <utility>
+
+#include "db_sta/dbNetwork.hh"
+#include "odb/db.h"
+#include "odb/dbTransform.h"
+
+namespace gui {
+
+PinDensityDataSource::PinDensityDataSource(utl::Logger* logger)
+    : RealValueHeatMapDataSource(logger,
+                                 "pins",
+                                 "Pin Density",
+                                 "Pin",
+                                 "PinDensity")
+{
+}
+
+bool PinDensityDataSource::populateMap()
+{
+  if (getBlock() == nullptr) {
+    return false;
+  }
+
+  // Iterate through blocks hierarchically to gather the flattened data
+  // for this view.
+  std::vector<std::pair<odb::dbBlock*, odb::dbTransform>> blocks
+      = {{getBlock(), odb::dbTransform()}};
+
+  while (!blocks.empty()) {
+    auto [current_block, current_transform] = blocks.back();
+    blocks.pop_back();
+
+    const bool has_child_blocks = !current_block->getChildren().empty();
+
+    for (auto* inst : current_block->getInsts()) {
+      if (!inst->getPlacementStatus().isPlaced()) {
+        continue;
+      }
+
+      odb::dbMaster* master = inst->getMaster();
+      if (has_child_blocks) {
+        auto child = inst->getBlock()->findChild(master->getName().c_str());
```

Comment:
is this correct? I realize we dont have any open examples, but looking it up by name feels a little clunky (not for this PR).